auv_navigation/auv_navigation/scripts/follow_path_action_server.py

```diff
@@ -0,0 +1,233 @@
+                    self.source_frame,
+                    req.target_frame,
+                    req.angle_offset,
+                    req.keep_orientation
+                )
+                if path and path.poses:
+                    response.success = True
+                    response.message = "Valid path created"
+                    goal.path = path
+                else:
+                    rospy.logwarn("Path creation failed")
+                    response.success = False
+                    response.message = "Path creation failed. Starting direct alignment"
+                    
+            except Exception as e:
+                rospy.logerr(f"Error while creating path: {e}")
+        else:
+            response.success = True
+            response.message = "Won't create path. Starting direct alignment"
+        
+        self.action_client.send_goal(goal)
+        rospy.logdebug("Goal sent to the Action Server successfully.")
+        return response
+
+    def do_alignment(self, path=None, target_frame=None, angle_offset=0.0, keep_orientation=False):
+        """
+        This is a blocking loop that won't exit until alignment is done or preempted.
+        """
+        
+        if not path and not target_frame:
+            rospy.logerr("Both path and target_frame are None. Cannot perform alignment.")
+            return False
+            
+        try:
+            feedback = FollowPathFeedback()
+            
+            while not rospy.is_shutdown(): # try to succeed until rospy is shutdown
+                
+                if self.server.is_preempt_requested():
+                    stop_cmd = Twist()
+                    self.cmd_vel_pub.publish(stop_cmd)
+                    self.server.set_preempted() 
+                    rospy.logwarn("Alignment preempted (do_alignment)")
+                    return False
+                
+                self.frame_controller.enable_alignment()
+                # 1. If path is valid, do path-based alignment
+                if path and path.poses:
+                    self.path_pub.publish(path)
+                    current_pose = get_current_pose(self.tf_buffer, self.source_frame)
+                    if current_pose is None: 
+                        rospy.logwarn("Failed to get current pose. Retrying...")
+                        self.control_rate.sleep()
+                        continue
+                    
+                    carrot_pose = calculate_carrot_pose(path, current_pose, self.carrot_distance)
+                    broadcast_carrot_frame(self.tf_broadcaster, self.tf_buffer, self.source_frame, carrot_pose)
+                    
+                    trans_error, rot_error = self.frame_controller.get_error(
+                        self.tf_buffer,
+                        self.source_frame,
+                        self.carrot_frame
+                        # We don't need to set angle_offset or keep_orientation here since they are taken cared of in path. 
```

Comment:
In this case, get error gets the error between robot and carrot. In direct alignment, we also add keep_orientation and angle_offset here, since we don't have the path anymore. Notice how the legacy align frame controller applied angle_offset when getting transform (get_transform is the ancestor of get_error):
```py
def get_transform(
        self, source_frame, target_frame, angle_offset, time=rospy.Time(0)
    ):
        try:
            # Get the current transform
            transform = self.tf_buffer.lookup_transform(
                target_frame, source_frame, time, rospy.Duration(2.0)
            )

            trans = (
                transform.transform.translation.x,
                transform.transform.translation.y,
                transform.transform.translation.z,
            )
            rot = (
                transform.transform.rotation.x,
                transform.transform.rotation.y,
                transform.transform.rotation.z,
                transform.transform.rotation.w,
            )

            return trans, rot

            # Apply the angle offset to the rotation
            roll, pitch, yaw = euler_from_quaternion(rot)
            yaw -= angle_offset  # Adjust yaw with angle offset
            yaw = angles.normalize_angle(yaw)
            new_rot = quaternion_from_euler(roll, pitch, yaw)
```